feat(builtin-gateways): add rules-based policies

kleinfreund opened the pull request:
You can use the custom [`test-meshgateway` entity](https://deploy-preview-2249--kuma-gui.netlify.app/gui/meshes/default/gateways/builtin/test-meshgateway/overview) to look at a "real enough" and not randomized example.

## Changes

**chore: mock MeshHTTPRoute and MeshGateway _rules**

**feat(builtin-gateways): add rules-based policies**

Update the built-in gateway detail view to show which rules from MeshHTTPRoutes and MeshTCPRoutes apply to which listeners and to which destination services they resolve.

Move the Dataplanes table into its own tab because selecting a listener in the new UI makes the content beneath it jump around.

Signed-off-by: Philipp Rudloff <philipp.rudloff@konghq.com>

Resolves #2208.

## Screenshots

![image](https://github.com/kumahq/kuma-gui/assets/5774638/2171aeaa-2418-4570-b532-748a54299e51)

johncowen commented:
Heads up incase: I just noticed we are probably going to tread on toes here between this one and https://github.com/kumahq/kuma-gui/pull/2247, did you want to try and get 2247 in first?

kleinfreund commented:
I’m going to open this up for review now. There are some items left to discuss with design mostly around the topic of information hierarchy in the Routes and Services part of the view. There’s quite a lot of relevant information that needs to be shown. The most tricky aspect is nicely representing multiple rules (i.e. the pair of matches and default) within a route card.

johncowen commented:
Did I see right that we aren't using nested routes for this? And if not how come we aren't? I guess there's a reason not to?

kleinfreund commented:
> Did I see right that we aren't using nested routes for this? And if not how come we aren't? I guess there's a reason not to?

That’s a bit of a vague question, but I’m guessing you mean the boundary between the view where a selected listener is in play and the one where there isn’t. I had it like that before and found no benefits to it (rather only complications) and changed it again.

johncowen commented:
> (rather only complications)

> And if not how come we aren't? I guess there's a reason not to?

i.e. Could you detail these complications for me?

kleinfreund commented:
> i.e. Could you detail these complications for me?

It’s just more cumbersome to use a RouterView as the interface for a statically resolved component. You get no prop checking and need a separate route. There’s no need for a separate route here.

kleinfreund commented:
> Please could you use nested routing for this.
> 
> There have been several PRs over the past few weeks moving areas that weren't using nested routing to use nested routing. Doing this fixed several bugs and made several UX improvements (not to mention the DX improvements)

If you can point out what can be improved by using nested routes here, I will consider it. For the time being, I’ll consider this suggestion one of technicality. I’m not interested in changing things around "just because".

johncowen commented:
The entire project uses nested routing to do this sort of thing, I don't see any reason to do anything different here.

At a general level, it prevents very many negative things as has been seen in moving the places where it wasn't being used to use it.

johncowen commented on src/app/data-planes/data/index.ts:
```diff
@@ -277,9 +288,8 @@ export const Rule = {
         const proxy = typeof item.proxyRule !== 'undefined'
           ? [{
-            ...Rule.fromObject(item.proxyRule as InspectBaseRule),
+            ...Rule.fromObject(item.proxyRule as InspectBaseRule, 'proxy'),
```
I would rather these methods always took a single argument, this is why previously the ruleType was applied here instead of inside `Rule`. Moreover if we are still adding the `type` here also we may aswell continue to add the `ruleType` here.

If it gets to a point where we absolutely must pass extra things in then fair enough, but I don't think we at that point.

kleinfreund replied:
There was a specific reason why I needed to move the addition of `ruleType` inside the transformer: I wanted a stricter type for `ruleType` so that autocompletion of the field knows of the three possible values. That meant that in order for `fromObject` to return a `Rule`, it needs to set one of the three allowed values (`''` not being one of them). And that can only happen inside the `fromObject` function.

Now, theoretically `type` could also be pulled into the function, but for `type` there is neither an advantage nor a forcing factor: it doesn’t come from `rule` and its resulting type will only be `string` anyway. While it would be nice to have a `PolicyTypeName` union type of all policy type names to make this more precise, it would require quite a bit more work to establish one.

johncowen replied:
`type` should always be string seeing as policies are pluggable.

So your reasoning here not to have `'' | 'to' | 'from' | 'proxy'` is incase you accidently search for a `ruleType=''`? (which would return none)

kleinfreund replied:
> So your reasoning here not to have `'' | 'to' | 'from' | 'proxy'` is incase you accidently search for a `ruleType=''`? (which would return none)

Why would I allow `''` in the first place? It’s not useful and I can easily avoid it by pulling the rule type stuff inside.

johncowen replied:
Well yeah but if ideally we want to stick to the single argument `fromObject` interface, it kinda depends what we prefer to keep. I'm not sure in what circumstance anyone would try and search ruleType by `''` even if Typescript allowed them to do that.

We could also have a private type in the middle to allow us to create a rule without a `ruleType` and then add the type where we did previously. Let me look/think about this a little more.

Btw this isn't a blocker.

johncowen replied:
What if we used `to` as a default instead of `` inside `Rule.FromObject`, then (as was the case previous to the changes here) we immediately overwrite it with the value from the API always. That way we get the types you are after and we keep the single argument `fromObject` API?

There are other ways to do this (we could have multiple `To/From/ProxyRule.fromObject` etc) but I think this is the simplest and I think it answers both concerns, wdyt?

kleinfreund replied:
I think that’d be acceptable. I’ve changed this now.

johncowen commented on src/app/data-planes/data/index.ts:
```diff
@@ -222,24 +223,36 @@ export const SidecarDataplane = {
+        // An omitted or empty hostnames list implies the wildcard hostname (meaning any hostnames apply).
+        hostnames: Array.isArray(conf.hostnames) && conf.hostnames.length > 0 ? conf.hostnames : ['*'],
+        rules: Array.isArray(conf.rules) ? conf.rules : [],
+      },
```
Isn't config the stuff we show in the UI? Shouldn't we avoid amending that? Or at least provide some sort of `raw` value that we don't touch to then display?

kleinfreund replied:
Good point. Seeing as we’re now accessing specific data from a `conf` object, using `raw` becomes the preferred name for this, I think. I’ve changed this now and replaced the few instances of accessing `config` elsewhere to `raw` (for `Rule`, not for other objects that might still use `config`; we should change those, too).

johncowen commented on src/app/gateways/components/RuleMatch.vue:
```diff
@@ -0,0 +1,99 @@
+          <dd>
+            <span class="key">Type:</span> <span>{{ param.type }}</span>
+            <span class="key">Name:</span> <span>{{ param.name }}</span>
+            <span class="key">Value:</span> <span>{{ param.value }}</span>
```
Tiny thing: What do you think of a nested `dl` here?

kleinfreund replied:
I briefly thought about it, and didn’t really have a clear picture about it. It feels a bit "spammy" to mark up everything as a definition list. I’m not sure how meaningful the notion of "this term is defined by another definition list" is.

johncowen commented on src/app/gateways/components/RuleMatch.vue:
```diff
@@ -0,0 +1,99 @@
+      <div
+        v-for="(param, index) in props.match.queryParams"
+        :key="index"
+      >
```
I know a `div` is fine within a `dl` in order to wrap the dt/dd combo. Not totally sure about nested divs any idea?

kleinfreund replied:
Ah, good point. I don’t know for sure, but in this case, we don’t need two levels anyway, so I removed one.

johncowen commented on src/app/gateways/views/BuiltinGatewayDetailView.vue:
```diff
@@ -1,242 +1,45 @@
+    :params="{
+      mesh: '',
+      gateway: '',
+      listener: '0',
```
I don't know if you saw my comment offline, but I think we can use a `hostname:port` combo for an identifier for these? wdyt?

kleinfreund replied:
I’m not entirely certain that is reliable. There is certainly validation in place preventing a clash of `${hostname}:${port}`, but I don’t know if there are edge cases that make this more tricky than it seems. While it’s not exactly pretty, I think the index is a reliable way to identify a listener for now (until someone adds an non-last listener, but the same problem exists when changing hostname or port).

Also, note that as far as I can tell, only the port is required. In the case of an absent hostname, the hostname is implied to be `*`.

Slightly related: I would like for the `/_rules` API to identify which listeners their rules apply to. This, too, would need to use _some_ form of identifier (even if it’s just the listener index).

johncowen replied:
Yeah I'm not 100% sure either, but considering there is validation stopping you from adding a clash of a `hostname:port` combo, I'd say its extremely unlikely we can't rely on it. Again considering there is validation preventing you from doing this, I'd almost say it would be a bug if someone did manage to do that.

Absent hostname would produce a `/*:80/` in the URL.

---

Separately, if we are using the URLs as state (ie. bookmarkable, sharable state), what happens if I share a link to listener `/3/` and then I delete listener number 2? Listener 3 would become listener 2 right? Then the link I just shared would be wrong.

kleinfreund replied:
> what happens if I share a link to listener `/3/` and then I delete listener number 2? Listener 3 would become listener 2 right?

Yes. But the link to whatever `${hostname}:${port}` would’ve been for listener 2 would also no longer work. 🤷🏻 

I’ll consider updating this, but I might do it in a follow up.

johncowen replied:
> But the link to whatever ${hostname}:${port} would’ve been for listener 2 would also no longer work.

And that would be the correct behaviour becuase I've deleted it. "Listener 3" would still be available at the same link of `hostname:port`.

kleinfreund replied:
For the sake of trying how this would work, I implemented this locally and there’s a problem with it: we don’t know `${hostname}:${port}` ahead of getting the MeshGateway. That means we can’t load the view with a pre-selected listener. We would need to update the URL after loading the MeshGateway. I’m going to leave this as an index for now.

johncowen reviewed (commented):
Additional to the nested routes thing, I did a quick scan here and added a few more comments/questions. Lemme know on those.

lahabana commented:
- Compact Name and Value for `matches` 
- Add tags for meshServiceSubset (represent the targetRef). Can we design a component for targetRef rendering/linking? --> Potentialy do as a follow up issue/PR
- For weight e.g: :wrench::20 can we have a weight icon and a tooltip to say `weight`. We can ask team to get an icon for this: https://kong.github.io/icons/

kleinfreund commented:
> 1. Compact Name and Value for `matches`
> 2. Add tags for meshServiceSubset (represent the targetRef). Can we design a component for targetRef rendering/linking? --> Potentialy do as a follow up issue/PR
> 3. For weight e.g: 🔧:20 can we have a weight icon and a tooltip to say `weight`. We can ask team to get an icon for this: https://kong.github.io/icons/
> 4. Show a `*` matcher when `matchers` is empty (similar to what we do for policies).
> 5. Remove `Overview and type` to just have `MeshHTTPRoute` or `MeshTCPRoute` in top left (like we do for listener type).
> 6. Needs support for filters. (how do this relate to point 8).
> 7. Discuss with design to not show matchers and origin by default (have a toggle for users to see them).
> 8. Discuss with design the possibility of having Matches and Services next to each other instead of one after the other.

I’ve done the following now:

> 1. Compact Name and Value for `matches`
> 2. Add tags for meshServiceSubset (represent the targetRef). Can we design a component for targetRef rendering/linking? --> Potentialy do as a follow up issue/PR
> 3. For weight e.g: 🔧:20 can we have a weight icon and a tooltip to say `weight`.
> 4. Show a `*` matcher when `matchers` is empty (similar to what we do for policies).
> 5. Remove `Overview and type` to just have `MeshHTTPRoute` or `MeshTCPRoute` in top left (like we do for listener type).

Left to do:

> 6. Needs support for filters. (how do this relate to point 8).
> 7. Discuss with design to not show matchers and origin by default (have a toggle for users to see them).
> 8. Discuss with design the possibility of having Matches and Services next to each other instead of one after the other.

... and the weight icon thing, but that can be done in a follow up, too.

lahabana commented:
Seems like we've lost the path type now:

<img width="1085" alt="image" src="https://github.com/kumahq/kuma-gui/assets/702355/72269c03-bfdb-4263-8464-081187cac1ab">

kleinfreund commented:
> Seems like we've lost the path type now:

I’ve fixed this.

I also implemented this now:

> 6\. Needs support for filters. (how do this relate to point 8).

This is a bit of a best effort implementation as there’s quite a bit to handle with filters. We can iterate on this in a follow up.

kleinfreund commented:
I created an issue for the two outstanding items that depend on design input we discussed previously: https://github.com/kumahq/kuma-gui/issues/2368

The rest of the feedback was addressed and this PR can be re-reviewed.

johncowen commented:
Is it just me or am I not seeing any listeners now on the preview site now? No matter which one I click on I get this:

<img width="751" alt="Screenshot 2024-04-02 at 12 31 45" src="https://github.com/kumahq/kuma-gui/assets/554604/c570433b-9fc6-4116-87f4-617e48703ab2">

kleinfreund commented:
> Is it just me or am I not seeing any listeners now on the preview site now? No matter which one I click on I get this:

I can definitely get gateways with listeners entries, but it does seem a little "too rare" for preview purposes. I can see if I can tweak this somehow.

johncowen commented:
Ok cool, if it's just a mock thing don't worry too much about it

johncowen commented on src/test-support/mocks/src/meshes/_/meshgateways.ts:
```diff
@@ -36,14 +38,27 @@ export default ({ env, fake, pager }: EndpointDependencies): MockResponder => (r
+              return {
+                hostname: `${fake.internet.domainWord()}.${fake.internet.domainName()}`,
+                port: fake.internet.port(),
+                protocol: fake.helpers.arrayElement<'HTTP' | 'HTTPS' | 'TCP' | 'TLS'>(['HTTP', 'HTTPS', 'TCP', 'TLS']),
```
Do we need to specify the types for these things? (I think I saw another one further up also)

kleinfreund replied:
Yes, the method is otherwise typed as returning an array of strings which fails the expected type.

johncowen replied:
What if we used `as const` on these instead?

kleinfreund replied:
I don’t see a reason to do that. Although a bit verbose, this gives us what we need here.

johncowen commented on src/test-support/FakeKuma.ts:
```diff
@@ -294,6 +295,43 @@ export class KumaModule {
+        ],
+      }),
+    }
+  }
```
I'd like to try to avoid having to many 'heavy' things on the fake kuma object. Smaller things aren't so bade (like `tags` for example). Here something like `kuma.httpHeaders` etc would be better than the full 'heavier' ruleMatch object.

Similarly to our actual application frontend templates, our mock templates should be made up of these smaller 'components' to give more flexibility/visibility - sometimes a bit of repetition is an ok trade for flexibility/visibility.

Tinier thing which might be a part-consequence of the above is that I'm not entirely sure whether we'd ever get two headers with the same name in an API response - not sure what would happen in real kuma if you tried to do that.

Anyway, bit of a comment more than anything, not a hard request unless you wanted to change something here yourself.

johncowen commented on src/app/gateways/components/RuleFilter.vue:
```diff
@@ -0,0 +1,145 @@
+.text-neutral {
+  color: $kui-color-text-neutral;
+}
+</style>
```
Would it make sense for this and RuleMatch go into the `rules` module?

kleinfreund replied:
Yes, I’m moving them provided the directory already exists.

kleinfreund replied:
Done.

johncowen reviewed (commented):
I had a couple more comments, but generally I think we should aim to get this in and circle back/iterate at a later date with some things like:

1. Nested routing
2. More stable `hostname:port` route segments
3. Possibly DataCollection for filtering/finding relationships

Lmk on the other comments and I think we are good to merge this (assuming @lahabana s comments are covered)

kleinfreund commented on src/app/gateways/components/RuleMatch.vue:
```diff
@@ -0,0 +1,105 @@
+
+        <dd>
+          <div class="list">
+            <span><span class="text-neutral">{{ t('http.api.property.type') }}:</span> {{ header.type ?? 'Exact' }}</span>
```
I wanted to do this, but held off for the time being because the source of data is nested _quite_ deep in the _rules API. Eventually, we should probably do this though.

johncowen replied:
Cool lets consider this in later PRs then 👍

kleinfreund commented:
> * Fix i18n problem (don't i18n keys)
> * Don't use a pill for type/action use `key:value`
> 
> <img alt="image" width="525" src="https://private-user-images.githubusercontent.com/702355/318827447-77591809-8346-4fad-93cc-5abfead3eeb1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwNzE4NjksIm5iZiI6MTcxMjA3MTU2OSwicGF0aCI6Ii83MDIzNTUvMzE4ODI3NDQ3LTc3NTkxODA5LTgzNDYtNGZhZC05M2NjLTVhYmZlYWQzZWViMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwMlQxNTI2MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNzI3ZmIzYWQ3MDg3NzNkYzQyMDZhMmViYmQxOGZiNzE3ODNkNDM2MGJmY2I1YTYxYjc0ZWQyMTM1M2I3MTg1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.xGgIZD9kHPEvqwkW2VO7meddYrBHdk1pEANYM4sSgt8">
> * For request mirror don't put the number in the service pill. Put it aside as `50%`

I’ve addressed these items now.

kleinfreund commented:
@johncowen Please finish reviewing this.

johncowen reviewed (commented):
I noticed that I could not see a TCP listener with Routes/Services and when I tried it out it looks like I don't see the service on a TCP listener/route?

![Screenshot 2024-04-04 at 11 09 25](https://github.com/kumahq/kuma-gui/assets/554604/acecd058-4404-47b9-a8a5-70a9fa343df3)

Could we take a look at that?

Apart from that I added a couple of more minor things inline. lmk

kleinfreund commented on src/app/gateways/views/BuiltinGatewayDataplanesView.vue:
```diff
@@ -0,0 +1,250 @@
+              :data="[meshGateway]"
+              :errors="[error]"
+              :loader="false"
+            >
```
I’ve changed this.

kleinfreund commented on src/test-support/mocks/src/meshes/_/meshgateways/_/_rules.ts:
```diff
@@ -0,0 +1,130 @@
+  const name = req.params.name as string
+
+  const haxProxyRuleOverride = env('KUMA_DATAPLANE_PROXY_RULE_ENABLED', '')
+  const haxProxyRule = haxProxyRuleOverride !== '' ? haxProxyRuleOverride === 'true' : fake.datatype.boolean()
```
Fixed.

kleinfreund commented on src/test-support/mocks/src/meshes/_/meshgateways/test-meshgateway/_.ts:
```diff
@@ -0,0 +1,51 @@
+      },
+    } satisfies MeshGateway,
+  }
+}
```
For PR previews for one and generally for testing with more real/less random data.

kleinfreund commented:
> Could we take a look at that?

That looks like an invalid TCP route. It has no `matches` property in its `rules` objects and so the `_rules` API doesn’t have any rules for this.

lahabana commented:
LGTM from a design POV

johncowen reviewed (approved):
Awesome, here's what I'm seeing with that applied 👍 

![Screenshot 2024-04-04 at 12 37 13](https://github.com/kumahq/kuma-gui/assets/554604/d0828d59-999c-4d64-a9a3-85a33f4a0370)

Lets get this in!
